parlai/agents/fairseq/fairseq.py

```diff
@@ -4,346 +4,434 @@
-        batch_reply = [{'id': self.getID()} for _ in range(bsz)]
-
-        # convert the observations into batches of inputs and targets
-        # valid_inds tells us the indices of all valid examples
-        # e.g. for input [{}, {'text': 'hello'}, {}, {}], valid_inds is [1]
-        # since the other three elements had no 'text' field
-
-        # also, split observations into sub-batches based on number of gpus
-        obs_split = np.array_split(observations, self.trainer.num_replicas)
-        samples = [self.batchify(obs) for obs in obs_split]
-        samples = [s for s in samples if s[0] is not None]
-        any_valid = any(len(s[0]) > 0 for s in samples)
-
-        if not any_valid:
-            # no valid examples, just return the empty responses we set up
+        batch_reply = [{"id": self.getID()} for _ in range(bsz)]
+
+        # torchagent boilerplate
+        self.is_training = any(["labels" in obs for obs in observations])
+        vec_obs = [self.vectorize(obs) for obs in observations]
+        xs, _, ys, _, valid_inds = self.map_valid(vec_obs)
+        if xs is None or ys is None:
+            print("WARNING: skipping this batch")
             return batch_reply
 
-        # produce predictions if testing; otherwise, train
-        has_targets = any(s[1] is not None for s in samples)
-        if not has_targets:
-            offset = 0
-            for s in samples:
-                xs = s[0]
-                valid_inds = s[2]
-
-                predictions = self._generate(self.args, xs)
-                for i in range(len(predictions)):
-                    # map the predictions back to non-empty examples in the batch
-                    batch_reply[valid_inds[i] + offset]['text'] = predictions[i]
-                    if i == 0:
-                        print('prediction:', predictions[i])
-                offset += len(valid_inds)
-        else:
-            loss = self._train(samples)
+        # here begins fairseq specific stuff
+        samples = self._make_sample(xs, ys)
 
-            batch_reply[0]['metrics'] = {}
-            for k, v in loss.items():
-                batch_reply[0]['metrics'][k] = v * bsz
+        if self.is_training:
+            self.model.train()
+            self.trainer.train_step(samples)
+        else:
+            # grade the evaluation label
+            self.model.eval()
+            self.trainer.valid_step(samples)
+
+            # Grade each of the candidate sequences
+            # TODO: grade everything in observations[i]['label_candidates']
+
+            # Next generate freely to create our response
+            if self.args.skip_generation:
+                # skip the generation step
+                for i in valid_inds:
+                    batch_reply[i]["text"] = ""
+            else:
+                # actually do the generation
+                for i, response in zip(valid_inds, self._generate(samples)):
+                    batch_reply[i]["text"] = response
```

Comment:
this setup looks like a good opportunity for torch agent batch_act to be generic and call e.g. `self.train_step` or `self.valid_step` and hoave those be overridable by the children, I think--with the rest being boilerplate.

Reply:
Agreed, this is one way we could improve the usability of torch_agent